Resubmit FSDP hybrid shard

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86722

Add a new sharding strategy `hybrid_shard` to FSDP. This sharding strategy shards across the node and replicates between nodes. This means that allreduce is done between nodes and _reduce_scatter is within nodes. This can help save communication volume as FSDP related collectives happen only within a node.

Implementation:
- hybrid_shard should be consistent across all FSDP modules contained within the root. Unlike other sharding strategies we don't support mixing and matching hybrid_shard yet.
- User is expected to pass in the appropriate inter-node process group
- FSDP internals generate the cross-node process groups. For example, if we have node0: 0, 1 node1: 0,1, node2: 0, 1, there will be 2 inter-node process groups: containing [(node0 local rank 0), (node1 local rank 0), ...] and so forth.

Differential Revision: [D40277048](https://our.internmc.facebook.com/intern/diff/D40277048/)

rohan-varma committed:
[WIP] Resubmit FSDP hybrid shard

Add a new sharding strategy `hybrid_shard` to FSDP. This sharding strategy shards across the node and replicates between nodes. This means that allreduce is done between nodes and _reduce_scatter is within nodes. This can help save communication volume as FSDP related collectives happen only within a node.

Implementation:
- hybrid_shard should be consistent across all FSDP modules contained within the root. Unlike other sharding strategies we don't support mixing and matching hybrid_shard yet.
- User is expected to pass in the appropriate inter-node process group
- FSDP internals generate the cross-node process groups. For example, if we have node0: 0, 1 node1: 0,1, node2: 0, 1, there will be 2 inter-node process groups: containing [(node0 local rank 0), (node1 local rank 0), ...] and so forth.

Differential Revision: [D40277048](https://our.internmc.facebook.com/intern/diff/D40277048/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Resubmit FSDP hybrid shard"


Add a new sharding strategy `hybrid_shard` to FSDP. This sharding strategy shards across the node and replicates between nodes. This means that allreduce is done between nodes and _reduce_scatter is within nodes. This can help save communication volume as FSDP related collectives happen only within a node.

Implementation:
- hybrid_shard should be consistent across all FSDP modules contained within the root. Unlike other sharding strategies we don't support mixing and matching hybrid_shard yet.
- User is expected to pass in the appropriate inter-node process group
- FSDP internals generate the cross-node process groups. For example, if we have node0: 0, 1 node1: 0,1, node2: 0, 1, there will be 2 inter-node process groups: containing [(node0 local rank 0), (node1 local rank 0), ...] and so forth.

Differential Revision: [D40277048](https://our.internmc.facebook.com/intern/diff/D40277048/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1858,6 +1871,35 @@ def _lazy_init(self) -> None:
+            sharding_backend = dist.get_backend(self.process_group)
+            # TODO - we are using the overall world size here, but user may wish
+            # to tune this.
+            world_size = dist.get_world_size()
```
Does this assume that FSDP's `self.process_group` is the default process group? (I.e. do we need to call `world_size = dist.get_world_size(self.process_group)`?)

rohan-varma replied:
This is a great catch, we should definitely pass in the process group

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1858,6 +1871,35 @@ def _lazy_init(self) -> None:
             self._init_param_attributes(handle)
         self._exec_order_data.init(self, self.process_group)
+        # Initialize inter-node process group for hybrid shard comm. We assume
+        # passed in process group is the process group over which the model is
```
> We assume passed in process group is the process group over which the model is sharded

Does "passed in process group" refer to the one passed to the FSDP constructor?

If so, I am wondering if this is really an assumption. It seems to be deterministically true given how FSDP sharding is implemented, and there would be a bug with FSDP if it somehow used a different process group that is not the one passed into the constructor.

awgu reviewed (commented):
I made an initial conceptual pass over the PR. I will make another pass later for nits after others have had a chance to take a look.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -151,15 +151,20 @@ class ShardingStrategy(Enum):
+    HYBRID_SHARD: Apply ``FULL_SHARD`` intra-node and
+                  ``NO_SHARD`` inter-node. Note that if this strategy is
+                  selected, user is expected to pass in the process group
+                  over which to shard over as ``process_group`` argument into
```
```suggestion
                  over which to shard as ``process_group`` argument into
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -842,11 +848,18 @@ class FullyShardedDataParallel(nn.Module):
             module to be wrapped with FSDP.
         process_group (Optional[ProcessGroup]):
-            process group for sharding
+            process group over which to shard wrapped modules over.
```
```suggestion
            process group over which to shard wrapped modules.
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -842,11 +848,18 @@ class FullyShardedDataParallel(nn.Module):
             Config sharding algorithm, different sharding algorithm has trade
             off between memory saving and communication overhead. ``FULL_SHARD``
-            will be chosen if sharding_strategy is not specified.
+            will be chosen if sharding_strategy is not specified. Note that if
```
```suggestion
            will be chosen if ``sharding_strategy`` is not specified. Note that if
```

awgu reviewed (commented):
Some nits for now.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+
+
+@contextlib.contextmanager
+def patch_allreduce(new_allreduce):
```
nit: Curious when to use `allreduce` vs. `all_reduce`.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_hybrid_shard_basic_setup(self):
+        """Tests that ``auto_wrap_policy`` propagates ``device_id`` to all
```
Update the test docstring?

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+        )
+
+
+        orig_ar = dist.all_reduce
```
nit: Maybe do not save this as a local variable at all, and directly call `dist.all_reduce()` in `patched_allreduce()` instead?

rohan-varma replied:
This might be an infinite recursion, right? If I directly call dist.all_reduce within the `patched_allreduce`, `patch_allreduce` will have overriden dist.all_reduce to point to `patched_aullreduce` so it would just call itself.

awgu replied:
Ah I see.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+        allreduce_called = False
+
+        def patched_allreduce(*args, **kwargs):
+            print(" -- patched --")
```
nit: I am assuming this is for local testing.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+            loss = fsdp_model.get_loss(inp, out)
+            loss.backward()
+
+        self.assertTrue(allreduce_called)
```
Should we change `allreduce_called` into a counter and verify the number of all-reduces is expected? If we go for that path, it will require some peaking into FSDP internals though.

```
num_flat_params = sum(len(fsdp_module._handles) for fsdp_module in FSDP.fsdp_modules(fsdp_model))
expected_num_all_reduces = num_flat_params  # one all-reduce per `FlatParameter`
```

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,123 @@
+            # 1 rank.
+            inter_node_pg = fsdp_module._inter_node_pg
+            my_rank = dist.get_rank(self.process_group)
+            other_ranks = [
```
nit: Unused variable.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -842,11 +848,18 @@ class FullyShardedDataParallel(nn.Module):
+            single node, which can be generated via ``dist.new_subgroups()``. In
+            addition, if ``HYBRID_SHARD`` is specified it must be consistent across
+            all FSDP instances that belong to the same root, i.e. mixing and matching
+            sharding strategies including ``HYBRID_SHARD`` is unsupported at the
```
Where is the exact point at which this breaks?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1858,6 +1871,35 @@ def _lazy_init(self) -> None:
+            world_size = dist.get_world_size()
+            # Assuming fully homogeneous setup
+            num_devices = torch.cuda.device_count()
+            num_distinct_nodes = world_size // num_devices
```
super nit:  `num_distinct_nodes` -> `num_nodes`? (basically, not sure if "distinct" matters)

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1871,6 +1913,13 @@ def _lazy_init(self) -> None:
                 )
                 fsdp_module._is_root = False
+                if self.sharding_strategy == ShardingStrategy.HYBRID_SHARD:
+                    assert hasattr(self, '_inter_node_pg')
```
Is this for mypy? Both attributes are set in this same method, so I am wondering if these asserts are meaningful.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3580,7 +3633,10 @@ def _cast_grad_to_param_dtype(
     def _should_free_unsharded_flat_param(self, handle: FlatParamHandle):
```
Before this PR: I wonder if we should add a comment that this method is meant for the post-backward reshard only.

awgu reviewed (commented):
More comments :)

rohan-varma committed:
Update on "Resubmit FSDP hybrid shard"


Add a new sharding strategy `hybrid_shard` to FSDP. This sharding strategy shards across the node and replicates between nodes. This means that allreduce is done between nodes and _reduce_scatter is within nodes. This can help save communication volume as FSDP related collectives happen only within a node.

Implementation:
- hybrid_shard should be consistent across all FSDP modules contained within the root. Unlike other sharding strategies we don't support mixing and matching hybrid_shard yet.
- User is expected to pass in the appropriate inter-node process group
- FSDP internals generate the cross-node process groups. For example, if we have node0: 0, 1 node1: 0,1, node2: 0, 1, there will be 2 inter-node process groups: containing [(node0 local rank 0), (node1 local rank 0), ...] and so forth.

Differential Revision: [D40277048](https://our.internmc.facebook.com/intern/diff/D40277048/)

[ghstack-poisoned]

zhaojuanmao commented:
possibly we still need to let users pass the global/default process group in the API? especially currently the inter node progress group is made based on assumption that intra node progress group is made for local host. could we spawn the intra node progress groups inside FSDP internally, so that users can use the API in the same way as using the default FSDP sharding strategy?

rohan-varma committed:
Update on "Resubmit FSDP hybrid shard"


Add a new sharding strategy `hybrid_shard` to FSDP. This sharding strategy shards across the node and replicates between nodes. This means that allreduce is done between nodes and _reduce_scatter is within nodes. This can help save communication volume as FSDP related collectives happen only within a node.

Implementation:
- hybrid_shard should be consistent across all FSDP modules contained within the root. Unlike other sharding strategies we don't support mixing and matching hybrid_shard yet.
- User is expected to pass in the appropriate inter-node process group
- FSDP internals generate the cross-node process groups. For example, if we have node0: 0, 1 node1: 0,1, node2: 0, 1, there will be 2 inter-node process groups: containing [(node0 local rank 0), (node1 local rank 0), ...] and so forth.

Differential Revision: [D40277048](https://our.internmc.facebook.com/intern/diff/D40277048/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1863,6 +1876,35 @@ def _lazy_init(self) -> None:
         for handle in self._handles:
             self._init_param_attributes(handle)
         self._exec_order_data.init(self, self.process_group)
+        # Initialize inter-node process group for hybrid shard comm. We assume
```
Just as my 2 cents that might help transition, to prepare for rebasing and/or after rebasing, we should
- refactor this inter-node process group initialization into its own helper method like `_init_inter_node_process_groups(state: _FSDPState)` that goes in `_init_utils.py`
- only call `_init_inter_node_process_groups()` in `FullyShardedDataParallel.__init__()` (i.e. no need for composable FSDP)

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3614,6 +3665,8 @@ def _post_backward_hook(
                     input_flattened = F.pad(grad_flatten, [0, num_pad])
                     output = torch.zeros_like(chunks[0])
                     self._communication_hook(self._communication_hook_state, input_flattened, output)
+                    if self.sharding_strategy == ShardingStrategy.HYBRID_SHARD:
```
(Easy) This can be added to `_post_backward_hook()` in `_runtime_utils.py`.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3712,7 +3765,10 @@ def _cast_grad_to_param_dtype(
     def _should_free_unsharded_flat_param(self, handle: FlatParamHandle):
```
This method is renamed to `_should_free_in_backward()` and is in `_runtime_utils.py`.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1863,6 +1876,35 @@ def _lazy_init(self) -> None:
+            sharding_backend = dist.get_backend(self.process_group)
+            # TODO - we are using the overall world size here, but user may wish
+            # to tune this.
+            world_size = dist.get_world_size(self.process_group)
```
I wanted to double check this line. In your implementation, `self.process_group` is the intra-node process group. Would `world_size` not always be `8`, when we want it to be the global world size?

(I checked the `dist.get_world_size()` docs, and it says that it returns the number of workers in the passed-in process group.)

awgu reviewed (commented):
I agree with @zhaojuanmao's comment. I think we should revisit the API.

My suggestion is to do the following:
1. If the user passes `None` for `process_group`, then we automatically construct the intra-node and inter-node process groups. This _requires_ that an auto wrap policy is passed.
2. If the user passes `None` for `process_group` and `auto_wrap_policy`, then we raise an error immediately.
3. If the user wants to perform manual wrapping, then the user must create the intra-node and inter-node process groups and pass them to the `process_group` argument as a `tuple`. We can expose a public helper to do this.

My motivation is so that (1) can cover the majority of cases and keep the usage simple, hiding the process group creation, while (3) can cover the remaining users employing manual wrapping.

If you do not have bandwidth and do not mind, I can revive this and try to finish.

awgu commented:
> Discussed with @awgu offline - @awgu has come up with a prototype for HSDP and I will help carry that forward with the discussions above.

The prototype lives at https://github.com/awgu/pytorch/tree/hsdp for reference.
